In [1]:
import pandas as pd
import numpy as np

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
ttc = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/MID TEST/Copy of titanic_passenger.csv')

In [5]:
ttc

Survived     Sex   Age  SiblingsSpouse  ParentChildren     Fare  \
0           0    male  22.0               1               0   7.2500   
1           1  female  38.0               1               0  71.2833   
2           1  female  26.0               0               0   7.9250   
3           1  female  35.0               1               0  53.1000   
4           0    male  35.0               0               0   8.0500   
..        ...     ...   ...             ...             ...      ...   
886         0    male  27.0               0               0  13.0000   
887         1  female  19.0               0               0  30.0000   
888         0  female   NaN               1               2  23.4500   
889         1    male  26.0               0               0  30.0000   
890         0    male  32.0               0               0   7.7500   

    EmbarkedCity   Class Deck  
0              S   Third  NaN  
1              C   First    C  
2              S   Third  NaN  
3              S   First    C  
4              S   Third  NaN  
..           ...     ...  ...  
886            S  Second  NaN  
887            S   First    B  
888            S   Third  NaN  
889            C   First    C  
890            Q   Third  NaN  

[891 rows x 9 columns]

In [6]:
ttc.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Survived        891 non-null    int64  
 1   Sex             891 non-null    object 
 2   Age             714 non-null    float64
 3   SiblingsSpouse  891 non-null    int64  
 4   ParentChildren  891 non-null    int64  
 5   Fare            891 non-null    float64
 6   EmbarkedCity    889 non-null    object 
 7   Class           891 non-null    object 
 8   Deck            203 non-null    object 
dtypes: float64(2), int64(3), object(4)
memory usage: 62.8+ KB


DROP THE COLUMN

In [7]:
ttc.drop('Deck', axis = 'columns', inplace = True)

In [8]:
ttc.columns

Index(['Survived', 'Sex', 'Age', 'SiblingsSpouse', 'ParentChildren', 'Fare',
       'EmbarkedCity', 'Class'],
      dtype='object')

Kolom DECK sudah tidak ada

In [9]:
ttc.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Survived        891 non-null    int64  
 1   Sex             891 non-null    object 
 2   Age             714 non-null    float64
 3   SiblingsSpouse  891 non-null    int64  
 4   ParentChildren  891 non-null    int64  
 5   Fare            891 non-null    float64
 6   EmbarkedCity    889 non-null    object 
 7   Class           891 non-null    object 
dtypes: float64(2), int64(3), object(3)
memory usage: 55.8+ KB


Dari hasil pengecekan data didapatkan bahwa:
1. Range index data ini sebanyak 891 entries
2. Kolom Age, Embarked City masih terdapat missing value.
3. Kolom Deck memiliki missing value lebih dari 60% maka kolom tersebut dapat dihilangkan karena sulit untuk mendapakan metode imputasi yang akurat.


**REMOVE THE ROWS**

In [10]:
ttc

Survived     Sex   Age  SiblingsSpouse  ParentChildren     Fare  \
0           0    male  22.0               1               0   7.2500   
1           1  female  38.0               1               0  71.2833   
2           1  female  26.0               0               0   7.9250   
3           1  female  35.0               1               0  53.1000   
4           0    male  35.0               0               0   8.0500   
..        ...     ...   ...             ...             ...      ...   
886         0    male  27.0               0               0  13.0000   
887         1  female  19.0               0               0  30.0000   
888         0  female   NaN               1               2  23.4500   
889         1    male  26.0               0               0  30.0000   
890         0    male  32.0               0               0   7.7500   

    EmbarkedCity   Class  
0              S   Third  
1              C   First  
2              S   Third  
3              S   First  
4              S   Third  
..           ...     ...  
886            S  Second  
887            S   First  
888            S   Third  
889            C   First  
890            Q   Third  

[891 rows x 8 columns]

In [11]:
ttc_update = ttc.copy()
ttc_update.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Survived        891 non-null    int64  
 1   Sex             891 non-null    object 
 2   Age             714 non-null    float64
 3   SiblingsSpouse  891 non-null    int64  
 4   ParentChildren  891 non-null    int64  
 5   Fare            891 non-null    float64
 6   EmbarkedCity    889 non-null    object 
 7   Class           891 non-null    object 
dtypes: float64(2), int64(3), object(3)
memory usage: 55.8+ KB


In [12]:
ttc_update[ttc_update['EmbarkedCity'].isnull()]

Survived     Sex   Age  SiblingsSpouse  ParentChildren  Fare  \
61          1  female  38.0               0               0  80.0   
829         1  female  62.0               0               0  80.0   

    EmbarkedCity  Class  
61           NaN  First  
829          NaN  First

In [13]:
ttc_update[~ttc_update['EmbarkedCity'].isnull()]

Survived     Sex   Age  SiblingsSpouse  ParentChildren     Fare  \
0           0    male  22.0               1               0   7.2500   
1           1  female  38.0               1               0  71.2833   
2           1  female  26.0               0               0   7.9250   
3           1  female  35.0               1               0  53.1000   
4           0    male  35.0               0               0   8.0500   
..        ...     ...   ...             ...             ...      ...   
886         0    male  27.0               0               0  13.0000   
887         1  female  19.0               0               0  30.0000   
888         0  female   NaN               1               2  23.4500   
889         1    male  26.0               0               0  30.0000   
890         0    male  32.0               0               0   7.7500   

    EmbarkedCity   Class  
0              S   Third  
1              C   First  
2              S   Third  
3              S   First  
4              S   Third  
..           ...     ...  
886            S  Second  
887            S   First  
888            S   Third  
889            C   First  
890            Q   Third  

[889 rows x 8 columns]

In [14]:
### memilih baris yang tidak kosong
ttc_update = ttc_update[~ttc_update['EmbarkedCity'].isnull()].reset_index(drop = True)

In [15]:
ttc_update.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 889 entries, 0 to 888
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Survived        889 non-null    int64  
 1   Sex             889 non-null    object 
 2   Age             712 non-null    float64
 3   SiblingsSpouse  889 non-null    int64  
 4   ParentChildren  889 non-null    int64  
 5   Fare            889 non-null    float64
 6   EmbarkedCity    889 non-null    object 
 7   Class           889 non-null    object 
dtypes: float64(2), int64(3), object(3)
memory usage: 55.7+ KB


Kolom Embarked sudah dihilangkan missing valuesnya. Sehingga tersisa 889 rows yang sudah tidak ada missing valuesnya lagi. Jadi, kolom yang sisa adalah 'Age'.

In [16]:
ttc_update[ttc_update['Age'].isnull()]

Survived     Sex  Age  SiblingsSpouse  ParentChildren     Fare  \
5           0    male  NaN               0               0   8.4583   
17          1    male  NaN               0               0  13.0000   
19          1  female  NaN               0               0   7.2250   
26          0    male  NaN               0               0   7.2250   
28          1  female  NaN               0               0   7.8792   
..        ...     ...  ...             ...             ...      ...   
857         0    male  NaN               0               0   7.2292   
861         0  female  NaN               8               2  69.5500   
866         0    male  NaN               0               0   9.5000   
876         0    male  NaN               0               0   7.8958   
886         0  female  NaN               1               2  23.4500   

    EmbarkedCity   Class  
5              Q   Third  
17             S  Second  
19             C   Third  
26             C   Third  
28             Q   Third  
..           ...     ...  
857            C   Third  
861            S   Third  
866            S   Third  
876            S   Third  
886            S   Third  

[177 rows x 8 columns]

Berarti, terdapat sekitar 20% baris yang mengalami missing value di kolom Age.


cek median umur dari passenger di Titanic, berdasarkan tingkat keselamatan, jenis kelamin, dan kota keberangkatan

In [17]:
ttc_update.groupby(['Survived', 'Sex', 'EmbarkedCity'])['Age'].median()

Survived  Sex     EmbarkedCity
0         female  C               17.50
                  Q               30.50
                  S               25.00
          male    C               30.00
                  Q               31.00
                  S               28.25
1         female  C               27.00
                  Q               19.00
                  S               29.00
          male    C               26.50
                  Q               29.00
                  S               28.00
Name: Age, dtype: float64

In [18]:
ttc_update.groupby(['Survived', 'Sex', 'EmbarkedCity'])['Age'].median().reset_index()

Survived     Sex EmbarkedCity    Age
0          0  female            C  17.50
1          0  female            Q  30.50
2          0  female            S  25.00
3          0    male            C  30.00
4          0    male            Q  31.00
5          0    male            S  28.25
6          1  female            C  27.00
7          1  female            Q  19.00
8          1  female            S  29.00
9          1    male            C  26.50
10         1    male            Q  29.00
11         1    male            S  28.00

In [19]:
median_umur = ttc_update.groupby(['Survived', 'Sex', 'EmbarkedCity'])['Age'].median().reset_index()
median_umur

Survived     Sex EmbarkedCity    Age
0          0  female            C  17.50
1          0  female            Q  30.50
2          0  female            S  25.00
3          0    male            C  30.00
4          0    male            Q  31.00
5          0    male            S  28.25
6          1  female            C  27.00
7          1  female            Q  19.00
8          1  female            S  29.00
9          1    male            C  26.50
10         1    male            Q  29.00
11         1    male            S  28.00

In [20]:
### Ganti Nama Kolom 

median_umur['MedianAge'] = median_umur['Age']

In [21]:
median_umur

Survived     Sex EmbarkedCity    Age  MedianAge
0          0  female            C  17.50      17.50
1          0  female            Q  30.50      30.50
2          0  female            S  25.00      25.00
3          0    male            C  30.00      30.00
4          0    male            Q  31.00      31.00
5          0    male            S  28.25      28.25
6          1  female            C  27.00      27.00
7          1  female            Q  19.00      19.00
8          1  female            S  29.00      29.00
9          1    male            C  26.50      26.50
10         1    male            Q  29.00      29.00
11         1    male            S  28.00      28.00

In [22]:
median_umur.drop(['Age'], axis = 'columns', inplace = True)
median_umur

Survived     Sex EmbarkedCity  MedianAge
0          0  female            C      17.50
1          0  female            Q      30.50
2          0  female            S      25.00
3          0    male            C      30.00
4          0    male            Q      31.00
5          0    male            S      28.25
6          1  female            C      27.00
7          1  female            Q      19.00
8          1  female            S      29.00
9          1    male            C      26.50
10         1    male            Q      29.00
11         1    male            S      28.00

Mengisi missing value dengan median umur yang sudh dicari

In [23]:
ttc_update[ttc_update['Age'].isnull()].head(3)

Survived     Sex  Age  SiblingsSpouse  ParentChildren     Fare  \
5          0    male  NaN               0               0   8.4583   
17         1    male  NaN               0               0  13.0000   
19         1  female  NaN               0               0   7.2250   

   EmbarkedCity   Class  
5             Q   Third  
17            S  Second  
19            C   Third

In [25]:
median_umur

Survived     Sex EmbarkedCity  MedianAge
0          0  female            C      17.50
1          0  female            Q      30.50
2          0  female            S      25.00
3          0    male            C      30.00
4          0    male            Q      31.00
5          0    male            S      28.25
6          1  female            C      27.00
7          1  female            Q      19.00
8          1  female            S      29.00
9          1    male            C      26.50
10         1    male            Q      29.00
11         1    male            S      28.00

Dapat dilakukan dengan MERGE 
Merge TTC UPDATE dengan MEDIAN_UMUR

In [26]:
ttc_update.merge(median_umur, how = 'left')

Survived     Sex   Age  SiblingsSpouse  ParentChildren     Fare  \
0           0    male  22.0               1               0   7.2500   
1           1  female  38.0               1               0  71.2833   
2           1  female  26.0               0               0   7.9250   
3           1  female  35.0               1               0  53.1000   
4           0    male  35.0               0               0   8.0500   
..        ...     ...   ...             ...             ...      ...   
884         0    male  27.0               0               0  13.0000   
885         1  female  19.0               0               0  30.0000   
886         0  female   NaN               1               2  23.4500   
887         1    male  26.0               0               0  30.0000   
888         0    male  32.0               0               0   7.7500   

    EmbarkedCity   Class  MedianAge  
0              S   Third      28.25  
1              C   First      27.00  
2              S   Third      29.00  
3              S   First      29.00  
4              S   Third      28.25  
..           ...     ...        ...  
884            S  Second      28.25  
885            S   First      29.00  
886            S   Third      25.00  
887            C   First      26.50  
888            Q   Third      31.00  

[889 rows x 9 columns]

In [27]:
ttc_update = ttc_update.merge(median_umur, how = 'left')

In [28]:
ttc_update.head(6)

Survived     Sex   Age  SiblingsSpouse  ParentChildren     Fare  \
0         0    male  22.0               1               0   7.2500   
1         1  female  38.0               1               0  71.2833   
2         1  female  26.0               0               0   7.9250   
3         1  female  35.0               1               0  53.1000   
4         0    male  35.0               0               0   8.0500   
5         0    male   NaN               0               0   8.4583   

  EmbarkedCity  Class  MedianAge  
0            S  Third      28.25  
1            C  First      27.00  
2            S  Third      29.00  
3            S  First      29.00  
4            S  Third      28.25  
5            Q  Third      31.00

Terdapat kolom baru di ttc_update. Cek kembali ttc_update yang mengalami missing value pada kolom AGE


In [29]:
ttc_update[ttc_update['Age'].isnull()].head(3)

Survived     Sex  Age  SiblingsSpouse  ParentChildren     Fare  \
5          0    male  NaN               0               0   8.4583   
17         1    male  NaN               0               0  13.0000   
19         1  female  NaN               0               0   7.2250   

   EmbarkedCity   Class  MedianAge  
5             Q   Third       31.0  
17            S  Second       28.0  
19            C   Third       27.0

Setiap nilai nan di kolom AGE dapat diganti dengan MedianAge untuk tahap selanjutnya

In [30]:
ttc_update['Age'] = ttc_update['Age'].fillna(ttc_update['MedianAge'])

In [31]:
ttc_update.head(6)

Survived     Sex   Age  SiblingsSpouse  ParentChildren     Fare  \
0         0    male  22.0               1               0   7.2500   
1         1  female  38.0               1               0  71.2833   
2         1  female  26.0               0               0   7.9250   
3         1  female  35.0               1               0  53.1000   
4         0    male  35.0               0               0   8.0500   
5         0    male  31.0               0               0   8.4583   

  EmbarkedCity  Class  MedianAge  
0            S  Third      28.25  
1            C  First      27.00  
2            S  Third      29.00  
3            S  First      29.00  
4            S  Third      28.25  
5            Q  Third      31.00

In [32]:
ttc_update

Survived     Sex   Age  SiblingsSpouse  ParentChildren     Fare  \
0           0    male  22.0               1               0   7.2500   
1           1  female  38.0               1               0  71.2833   
2           1  female  26.0               0               0   7.9250   
3           1  female  35.0               1               0  53.1000   
4           0    male  35.0               0               0   8.0500   
..        ...     ...   ...             ...             ...      ...   
884         0    male  27.0               0               0  13.0000   
885         1  female  19.0               0               0  30.0000   
886         0  female  25.0               1               2  23.4500   
887         1    male  26.0               0               0  30.0000   
888         0    male  32.0               0               0   7.7500   

    EmbarkedCity   Class  MedianAge  
0              S   Third      28.25  
1              C   First      27.00  
2              S   Third      29.00  
3              S   First      29.00  
4              S   Third      28.25  
..           ...     ...        ...  
884            S  Second      28.25  
885            S   First      29.00  
886            S   Third      25.00  
887            C   First      26.50  
888            Q   Third      31.00  

[889 rows x 9 columns]

In [34]:
ttc_update.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 889 entries, 0 to 888
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Survived        889 non-null    int64  
 1   Sex             889 non-null    object 
 2   Age             889 non-null    float64
 3   SiblingsSpouse  889 non-null    int64  
 4   ParentChildren  889 non-null    int64  
 5   Fare            889 non-null    float64
 6   EmbarkedCity    889 non-null    object 
 7   Class           889 non-null    object 
 8   MedianAge       889 non-null    float64
dtypes: float64(3), int64(3), object(3)
memory usage: 69.5+ KB


Sudah tidak ada missing values di kolom Age